atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/sweep/GetCellTimestamps.java

```diff
@@ -0,0 +1,106 @@
+/*
+ * Copyright 2017 Palantir Technologies, Inc. All rights reserved.
+ *
+ * Licensed under the BSD-3 License (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://opensource.org/licenses/BSD-3-Clause
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.keyvalue.cassandra.sweep;
+
+import java.util.Collection;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import com.google.common.collect.Iterables;
+import com.google.common.collect.Lists;
+import com.palantir.atlasdb.keyvalue.api.Cell;
+import com.palantir.atlasdb.keyvalue.api.TableReference;
+import com.palantir.atlasdb.keyvalue.cassandra.CqlExecutor;
+
+public class GetCellTimestamps {
+
+    private final CqlExecutor cqlExecutor;
+    private final TableReference tableRef;
+    private final byte[] startRowInclusive;
+    private final int batchHint;
+
+    private final Collection<CellWithTimestamp> timestamps = Lists.newArrayList();
+
+    public GetCellTimestamps(
+            CqlExecutor cqlExecutor,
+            TableReference tableRef,
+            byte[] startRowInclusive,
+            int batchHint) {
+        this.cqlExecutor = cqlExecutor;
+        this.tableRef = tableRef;
+        this.startRowInclusive = startRowInclusive;
+        this.batchHint = batchHint;
+    }
+
+    /**
+     * Fetches a batch of timestamps, grouped by cell. The returned {@link CellWithTimestamp}s are ordered by cell.
+     */
+    public List<CellWithTimestamps> execute() {
+        fetchBatchOfTimestamps();
+
+        return groupTimestampsByCell();
+    }
+
+    /**
+     * We always finish the last whole row when fetching timestamps. Sweep actually only requires that we return a
```

Comment:
What if the last row is wide and does not fit in memory? Seems like this implementation is not OOM-proof. You need to page through the remainder of the last row as well.

Reply:
We did this previously with AllCellsPerRowPager, so I left the bahvior as is

Reply:
Fair point, however the implementation that you reverted avoided that.